test/machi_ap_repair_eqc.erl

```diff
@@ -0,0 +1,568 @@
+                            {undefined, undefined}
+                end,
+                pretty_commands(
+                  ?MODULE, Cmds, {Seq, Par, Res},
+                  aggregate(with_title(cmds), command_names(Cmds),
+                  collect(with_title(length5), (CmdsLen div 5) * 5,
+                  collect(with_title(conc),    concurrency(Cmds),
+                          {Dataloss, Critical} =:= {0, 0})))
+                 )
+            end)).
+
+%% Initilization / setup
+
+%% Fake initialization function for debugging in shell like:
+%% > eqc_gen:sample(eqc_statem:commands(machi_ap_repair_eqc)).
+%% but not so helpful.
+initial_state() ->
+    #state{cr_list=[a,b,c]}.
+
+initial_state(Num, Seed, Verbose) ->
+    AllListE = all_list_extra(Num),
+    FLUNames = [P#p_srvr.name || {P, _Dir} <- AllListE],
+    MgrNames = [{Name, machi_flu_psup:make_mgr_supname(Name)} || Name <- FLUNames],
+    Dict = orddict:from_list([{P#p_srvr.name, P} || {P, _Dir} <- AllListE]),
+
+    FCList = [begin
+                  {ok, PCPid} = machi_proxy_flu1_client:start_link(P),
+                  {Name, PCPid}
+              end || {_, #p_srvr{name=Name}=P} <- Dict],
+    %% CR clients are pooled, each has "name" which is interpreted "From"
+    %% side of simulated partition.
+    CRListCount = 10, % ad-hoc
+    SimSelfNames = lists:append(lists:duplicate(CRListCount div Num +1, FLUNames)),
+    CRList = [begin
+                  {ok, C} = machi_cr_client:start_link(
+                              [P || {_, P} <- Dict],
+                              [{use_partition_simulator, true},
+                               {simulator_self_name, SimSelfName},
+                               {simulator_members, FLUNames}]),
+                  {SimSelfName, C}
+              end || SimSelfName <- SimSelfNames],
+    #state{num=Num, seed=Seed, verbose=Verbose,
+           flu_names=FLUNames, mgr_names=MgrNames,
+           cr_list=CRList, fc_list=FCList}.
+
+setup_chain(Num, Seed, Verbose) ->
+    %% ?V("setup_chain(Num=~w, Seed=~w~nn", [Num, Seed]),
+    AllListE = all_list_extra(Num),
+    FLUNames = [P#p_srvr.name || {P, _Dir} <- AllListE],
+    MgrNames = [{Name, machi_flu_psup:make_mgr_supname(Name)} || Name <- FLUNames],
+    Dict = orddict:from_list([{P#p_srvr.name, P} || {P, _Dir} <- AllListE]),
+
+    ok = shutdown_hard(),
+    [begin
+         machi_flu1_test:clean_up_data_dir(Dir),
+         filelib:ensure_dir(Dir ++ "/not-used")
+     end || {_P, Dir} <- AllListE],
+    [catch ets:delete(T) || T <- tabs()],
+
+    [ets:new(T, [set, public, named_table,
+                 {write_concurrency, true}, {read_concurrency, true}]) ||
+        T <- tabs()],
+    {ok, _} = application:ensure_all_started(machi),
+
+    SimSpec = {part_sim,
+               {machi_partition_simulator, start_link, [{0,0,0}, 0, 100]},
+               permanent, 500, worker, []},
+    {ok, _PSimPid} = supervisor:start_child(machi_sup, SimSpec),
```

Comment:
In `shutdown_hard()`, application `machi` is stopped, which includes partition simulator process.